File: Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/dialogs/sso/ssoSignInDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ComponentDialog, WaterfallDialog, OAuthPrompt } = require('botbuilder-dialogs');

const SSO_SIGNIN_DIALOG = 'SsoSignInDialog';
const OAUTH_PROMPT = 'OAuthPrompt';
const WATERFALL_DIALOG = 'WaterfallDialog';

// Helps prepare the host for SSO operations and provides helpers to check the status and invoke the skill.
class SsoSignInDialog extends ComponentDialog {
    constructor(connectionName) {
        super(SSO_SIGNIN_DIALOG);

        this.addDialog(new OAuthPrompt(OAUTH_PROMPT, new OAuthPromptSettings({
            connectionName: connectionName,
            text: 'Sign in to the host bot using AAD for SSO',
            title: 'Sign In'
        })));

        this.addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
            this.signInStep.bind(this),
            this.displayTokenStep.bind(this)
        ]));

        this.initialDialogId = WATERFALL_DIALOG;
    }

    async signInStep(stepContext) {
        return await stepContext.beginDialog(OAUTH_PROMPT);
    }

    async displayTokenStep(stepContext) {
        if (typeof stepContext.result !== 'TokenResponse') {
            await stepContext.context.sendActivity('No token was provided.');
        } else {
            await stepContext.context.sendActivity(`Here is your token: ${ result.Token }`);
        }
        return await stepContext.endDialog();
    }
}

module.exports.SsoSignInDialog = SsoSignInDialog;

File: Bots/Python/Consumers/CodeFirst/WaterfallHostBot/dialogs/sso/sso_signin_dialog.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.dialogs import (
    ComponentDialog,
    WaterfallDialog,
    WaterfallStepContext,
    OAuthPrompt,
    OAuthPromptSettings,
)


class SsoSignInDialog(ComponentDialog):
    def __init__(self, connection_name: str):
        super().__init__(SsoSignInDialog.__name__)

        self.add_dialog(
            OAuthPrompt(
                OAuthPrompt.__name__,
                OAuthPromptSettings(
                    connection_name=connection_name,
                    text="Sign in to the host bot using AAD for SSO",
                    title="Sign In",
                ),
            )
        )

        self.add_dialog(
            WaterfallDialog(
                WaterfallDialog.__name__,
                [
                    self.signin_step,
                    self.display_token,
                ],
            )
        )

        self.initial_dialog_id = WaterfallDialog.__name__

    async def signin_step(self, step_context: WaterfallStepContext):
        return await step_context.begin_dialog(OAuthPrompt.__name__)

    async def display_token(self, step_context: WaterfallStepContext):
        if not step_context.result.token:
            await step_context.context.send_activity("No token was provided.")
        else:
            await step_context.context.send_activity(
                f"Here is your token: {step_context.result.token}"
            )
        return await step_context.end_dialog()
